src/Compiler/test/Microsoft.NET.Sdk.Razor.SourceGenerators.Tests/RazorSourceGeneratorTests.cs

```diff
@@ -1450,6 +1450,57 @@ async Task RenderMyRazor()
             Assert.Single(result.GeneratedSources);
         }
 
+        [Fact, WorkItem("https://github.com/dotnet/razor/issues/7049")]
+        public async Task SourceGenerator_CshtmlFiles_TagHelperInFunction_ManualSuppression()
+        {
+            // Arrange
+            var project = CreateTestProject(new()
+            {
+                ["Pages/Index.cshtml"] = """
+                @addTagHelper *, TestProject
+
+                @{ await RenderMyRazor(); }
+
+                @functions {
+                    #pragma warning disable 1998
+                    async Task RenderMyRazor()
+                    {
+                        var lambdaWithUnnecessaryAsync1 = async () => { };
+                        <email>first tag helper</email>
+                        <email>
+                            second tag helper
+                            <email>nested tag helper</email>
+                        </email>
+                        var lambdaWithUnnecessaryAsync2 = async () => { };
+                    }
+                }
+                """,
+            }, new()
+            {
+                ["EmailTagHelper.cs"] = """
+                using Microsoft.AspNetCore.Razor.TagHelpers;
+
+                public class EmailTagHelper : TagHelper
+                {
+                    public override void Process(TagHelperContext context, TagHelperOutput output)
+                    {
+                        output.TagName = "a";
+                    }
+                }
+                """
+            });
+            var compilation = await project.GetCompilationAsync();
+            var driver = await GetDriverAsync(project);
+
+            // Act
+            var result = RunGenerator(compilation!, ref driver,
+                d => Assert.Equal("Pages/Index.cshtml(15,52): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.", d.ToString()));
```

Comment:
Added a test to demonstrate that this still behaves differently (when user manually writes `#pragma warning disable 1998` in a `@functions` block). Prior to https://github.com/dotnet/razor/pull/8346, this warning would not be raised. If that's a problem, we can completely revert the original PR https://github.com/dotnet/razor/pull/8346 instead.

Reply:
This seems like a problem. The user disabled 1998, we should not be warning about it.

Reply:
It seems to me that `#pragma warning disable`s in Razor files will often fail to work since the Razor compiler emits them itself without considering user's code (so if user disables a warning which the Razor compiler then silently disables/restores, the warning gets enabled). This is just another case. But I can revert the original PR to mitigate the regression faster for now.

Reply:
@jaredpar do you have opinions? The options are to have this inconsistent behavior, or have https://github.com/dotnet/razor/issues/7049

Reply:
I don't like this behavior either. Would we consider reverting #8346 and fixing #8344 instead as a follow up? I know there's some worry about exceptions changing, but it feels unlikely to me that it would actually break people.

This is currently blocking aspnetcore's SDK update.

Reply:
@halter73 yes, I think that would be my preferred path as well.

Reply:
https://github.com/dotnet/razor/pull/8719